R/allGenerics.R

```diff
@@ -106,24 +150,92 @@ setMethod("[", signature(x = "TreeSummarizedExperiment"),
               } else {
                   nlc <- lc
               }
-              
-              
+
+              #
+              refSeq <- referenceSeq(x)
+              if (!missing(i)) {
+                  refSeq <- referenceSeq(x)
+                  if(!is.null(refSeq)){
+                      if(is(refSeq,"DNAStringSetList")){
+                          ii <- rep(list(i),length(refSeq))
+                          refSeq <- refSeq[ii]
+                      } else {
+                          refSeq <- refSeq[i]
+                      }
+                  }
+              }
+
               # Subset the traditional slots from SummarizedExperiment
               nx <- callNextMethod()
-              
+
               # update slots
               final <- BiocGenerics:::replaceSlots(nx,
                                                    rowLinks = nlr,
-                                                   colLinks = nlc)
-              
+                                                   colLinks = nlc,
+                                                   referenceSeq = refSeq)
+              validObject(final)
               return(final)
-          })
+          }
+)
+
+#' @importFrom methods callNextMethod
+#' @import SingleCellExperiment
+#' @importFrom S4Vectors metadata
+#' @rdname TreeSummarizedExperiment-accessor
+#' @export
+setReplaceMethod("[",
+    signature(x = "TreeSummarizedExperiment", "ANY", "ANY", 
+              "TreeSummarizedExperiment"),
+    function(x, i, j, ..., value){
+        x <- updateObject(x)
+        value <- updateObject(value)
+        if (missing(i) && missing(j)) {
+            # do callNextMethod because of objects potentially being updated
+            return(callNextMethod())
+        }
+
+
+        # TODO rowLinks
```

Comment:
So this is the part I am not sure about. Should links matching the selected lines in `value` be transfered to `x`? I would say no, since the `rowLinks` are not necessarily a 1:1 match.

If it is a no, nothing has to change in the function. Alternatively an argument `copy.links` or `drop.links` could be introduced to fine tune the behavior if you think that is necessary.

Reply:
Yes, that is what I mean. We can name the character vector that was put into  the `DNAStringSet` to track the match between rows and sequences

Reply:
That is not necessary. Here is an example:

```
# assays data
toyTable <- matrix(1:40, nrow = 10)
colnames(toyTable) <- paste(rep(LETTERS[1:2], each = 2), rep(1:2, 2), sep = "_")
rownames(toyTable) <- paste("entity", seq_len(10), sep = "")

data("tinyTree")

# row data
rowInf <- DataFrame(var1 = sample(letters[1:2], 10, replace = TRUE),
                    var2 = sample(c(TRUE, FALSE), 10, replace = TRUE),
                    row.names = rownames(toyTable))
# column data
colInf <- DataFrame(gg = c(1, 2, 3, 3),
                    group = rep(LETTERS[1:2], each = 2),
                    row.names = colnames(toyTable))
colTree <- rtree(4)
colTree$tip.label <- colnames(toyTable)
colTree$node.label <- c("All", "GroupA", "GroupB")

refSeq <- DNAStringSetList(one = DNAStringSet(rep("A",nrow(toyTable))),
                           two = DNAStringSet(rep("B",nrow(toyTable))))

tse <- TreeSummarizedExperiment(assays = list(toyTable),
                                rowData = rowInf,
                                rowNodeLab = tinyTree$tip.label,
                                colData = colInf,
                                rowTree = tinyTree,
                                colTree = colTree,
                                referenceSeq = refSeq)

# what should happen to the data in rowLinks()?
tse[1,] <- tse[2,]
```

Currently this last line throws an error due to rowLinks and the rownames being not inline.

Reply:
If the `rowLinks` is not a 1:1 match, we can set it like below, right?
For example, `a, b, c` are leaf labels 
```
x0 <- setNames(c("CTC-NACCAGTAT", "TTGA", "TACCTAGAG"),
               c("a", "a", "b"))
x1 <- setNames(c("TGA", "TGA", "TAC"),
               c("a", "a", "b"))


refSeq <- DNAStringSetList(one = DNAStringSet(x0),
                           two = DNAStringSet(x1))
refSeq
```

Reply:
This problem also is similar to `cbind` and `rbind` not working correctly on `TSE`. I think the grammar for these case is not "invented", yet. Maybe its a good time to do it now.

```
rbind(tse,tse)
cbind(tse,tse)
```

Reply:
> If the `rowLinks` is not a 1:1 match, we can set it like below, right?
> For example, `a, b, c` are leaf labels
> 
> ```
> x0 <- setNames(c("CTC-NACCAGTAT", "TTGA", "TACCTAGAG"),
>                c("a", "a", "b"))
> x1 <- setNames(c("TGA", "TGA", "TAC"),
>                c("a", "a", "b"))
> 
> 
> refSeq <- DNAStringSetList(one = DNAStringSet(x0),
>                            two = DNAStringSet(x1))
> refSeq
> DNAStringSetList of length 2
> [["one"]] a=CTC-NACCAGTAT a=TTGA b=TACCTAGAG
> [["two"]] a=TGA a=TGA b=TAC
> ```

The `DNAStringSet` is always expected to follow this contract `length(DNAStringSet) == nrow(x)`. The same also true for `DNAStringSetList`. In this case each element has to have the same length then the assay has rows.
Therefore the relationship is only 1:1, like for example `rowRanges`

Reply:
Still not sure how `1:1` would be an issue here. To clarify, the `rowLinks` and the row dimension of `assays` in `TSE`  are always `1:1` . `multiple : one`  is that multiple rows of `assays` can be mapped to one node of a `rowTree`. 

I am working on `rbind` and `cbind`. For `rbind`, one new column `whichTree` is added to the `rowLinks` to say about which tree of the `rowTree` the node is on. Does that would solve the issue here?

Reply:
For me the issue is not about the 1:1 of links to rows, but about the handling of the link in the situation. A few scenarios come to mind, which I think would all be valid somehow:

1. The links are not modified by `x[1,] <- x[2,]`. Only the data in the assays and the reference sequence is exchanged as defined here and elsewhere. This is the current situation and I think this is acceptable.
2. The links get exchanged, if the trees are the same. The user might expect the behaviour, if the first row is exchanged for the second row. Otherwise the relationship of data and tree wouldn't match anymore.
3. If the trees are not the same, than a few thinks could happen:
    - the tree gets pruned for the branches linking to the row being overwritten, if tip/branch is only linked to the one row. 
    - The same happens to the tree on the incoming object
    - both trees are joined at a "root"
4. Alternatively the trees do not get pruned and just merged at a "root". This would be the behavior similar to `rbind`
5. Add additional arguments to the `[<-` to let the user choose, which behaviour from the ones given above, should be used. For example `drop_links` could be used to select option 1.  

I acknowledge, that this is at first a non-issue since the current implementation is already among the 4 options I outlined, but a discussion might be helpful/needed, since with the introduction of the reference seq the `[<-` function needs to be implemented. Also maybe I missed an option in the list above.

And additional column `whichTree` might be helpful, if a merging of trees is not an option. To behonest I have never tried merging trees, so the options outlined above are just though constructs. However, the `bind.tree` function exist in the `ape` package, which maybe can be used for `rbind`/`cbind`.

Reply:
Or could you please give example cases when the `drop.links` and `copy.links` are required? It might help to see whether I should take it into consideration in the work of `rbind` and `cbind`.

Reply:
Regarding `cbind`/`rbind` three scenarios come to mind:

1. trees on both dimensions in both objects are present. The tree for the growing dimension can be merge and links adjusted. For the tree on the non-growing dimension the following options exist:
    - the tree and links from the first object persist. (default)
    - both trees on this dimension are removed: `drop.links = TRUE`
    - the trees are the same, but the links may not be. The links are copied from the last object: `copy.links = TRUE`
2. Only one tree exists on one of the objects for the non-growing dimension
    - the tree and links from the object, which has the data, persist. (default)
    - the tree(s) on this dimension is removed: `drop.links = TRUE`
3. Only one tree exists on one of the objects for the growing dimension
    - the tree(s) on this dimension is removed: `drop.links = TRUE`
    - new links are tried to be generated for the object not containing a tree, like when using `changeTree`. A warning gets issued if rows need to be dropped. (default)